### ISSUE TITLE
Profiles Compilers

### ISSUE THREAD

**Issue by jjscheel** [riscv-admin/dev-partners#16]
### Technical Group

Profiles TG

### ratification-pkg

Profiles

### Technical Liaison

Christoph Muellner

### Task Category

Compilers/Toolchain

### Task Sub Category

- [X] gcc
- [ ] binutils
- [ ] gdb
- [ ] intrinsics
- [ ] Java
- [ ] KVM
- [ ] ld
- [X] llvm
- [ ] Linux kernel
- [ ] QEMU
- [ ] Spike

### Ratification Target

1Q2023

### Statement of Work (SOW)

SOW: [link](https://docs.google.com/document/d/1Plbc-igDAudw62yYvdopFF4MbNKwpIxMRB7hzrq5_9M/)

**SOW Signoffs:** (delete those not needed)
  - Task group liaison sign-off date:
  - Development partner sign-off date:
  - Toolchains & Runtimes sign-off date (if gcc, LLVM, binutils, gdb work):

### Waiver

- [ ] Freeze
- [ ] Ratification

#### RISC-V PRs
- [RISC-V -march parameter documentation](https://github.com/riscv-non-isa/riscv-toolchain-conventions/pull/26)
- [RISC-V Profiles Conventions Proposal](https://github.com/riscv-non-isa/riscv-toolchain-conventions/pull/36)
...

#### Upstream PRs
- [gcc patch](https://gcc.gnu.org/pipermail/gcc-patches/2024-December/670667.html)

**Comment by jjscheel** [comment]
Mar 27 status from @cmuellner:
> Here is a short follow-up on this topic.
> We agreed to split up the proposal into four PRs to handle the changes
> more efficiently and to avoid confusion about what is current
> behaviour and what is targeted behaviour:
> * document current -march string
> * document -march string relaxation
> * add custom extensions
> * add profiles support
>
> Jiawei volunteered to drive this forward.
> Kito and I will support him as much as we can.

**Comment by jjscheel** [comment]
@pz9115, do you have a projected date by which this might be done and all PRs accepted?

Would you also provide links to the PRs here once they are submitted?

Thanks!

**Comment by pz9115** [comment]
I prepared this doc to record profile implement details in toolchain.

https://docs.google.com/document/d/1TZXRIgVfQHWQ6xrZflHXUCSav6xNmliojrW2bEsvPno/edit

**Comment by jjscheel** [comment]
@pz9115, thanks for the document link.  What is your projected date for all PRs accepted?

**Comment by pz9115** [comment]
Should I open a new PR for this proposal, I'm not sure.

**Comment by jjscheel** [comment]
@cmuellner, can you answer @pz9115's question?  I do not feel qualified to advise here.  Thanks!

**Comment by jjscheel** [comment]
@pz9115, per our discussion last week and lacking an answer from @cmuellner, I believe you should proceed with your PRs as you see best.  Christoph has clearly confirmed that you are in the lead here and documented what needs to be done.

Thanks for driving!

**Comment by cmuellner** [comment]
Sorry, I did not see the original notification back then.
Yes, sure, go ahead and create a new PR.

**Comment by pz9115** [comment]
Ok, I just open this new PR:
https://github.com/riscv-non-isa/riscv-toolchain-conventions/pull/36

**Comment by jjscheel** [comment]
@pz9115, how long do you think it will take to have all PRs written and accepted to the riscv repos?  Then, how long do you think it will take to get these changes upstreamed to the main communities?

I'm trying to pick a target date for when this item might be "done" (all PRs accepted upstream).  So, any thoughts on what would be a target date would be appreciated too.

**Comment by pz9115** [comment]
I think it may take one month for the PR merged, and may take another month for upstream works, hopefully it will faster than I thought.

**Comment by jjscheel** [comment]
Great.  How about we set a date for "Completion target" as July 31, 2023?

**Comment by pz9115** [comment]
Ok, for me, it seems need a long term.

**Comment by jjscheel** [comment]
I've updated the Completion Target date.  THANKS!!!

**Comment by cmuellner** [comment]
In today's SIG Toolchain call, the following reminder has been brought up: toolchain support for profiles does not only need a proper extension of the `-march`-string and a corresponding implementation in the toolchain components. It also requires adding support for the extensions that are ratified alongside the profiles (see chapter 7 of the ratified v1.0 profiles specification).

**Comment by jjscheel** [comment]
@pz9115, when you get a feel for how much work exists in these additional extensions, please provide some thoughts on when you expect to submit the final PR for this work.

**Comment by pz9115** [comment]
The first step of documenting the format of RISC-V Profiles  using in the compiler has been basically completed, and it will be officially merged into the riscv-toolchain-conventions document around the end of the month.
The second step is to support RISC-V profiles in the upstream of GCC. I had already sent an RFC patch, and it can be resent with some simple modifications. Of course, this will take a while to review, and it is estimated that there will be progress in an extra month.

**Comment by jjscheel** [comment]
Thanks, @pz9115.  Can you kindly provide links to your PRs as comments?  Then, I'll copy into the description so we have a nice summary when done.

**Comment by pz9115** [comment]
> Thanks, @pz9115. Can you kindly provide links to your PRs as comments? Then, I'll copy into the description so we have a nice summary when done.

Of course, thank you Jeff.

https://github.com/riscv-non-isa/riscv-toolchain-conventions/pull/36

**Comment by jjscheel** [comment]
@pz9115, thanks for the update.  

I notice that the PR is for the RISC-V repository.  We will want to track the work all the way into the upstream repositories for completeness.  So, please make sure to add those too when they are available (if you have them).  Thanks!

**Comment by pz9115** [comment]
Yes, It's for RISC-V repository and define the basic format that profile name should follow. It still under discussing about the Zicc* extensions effect for toolchain behaviour. Maybe recently will release a doc to make it more clearly.

**Comment by pz9115** [comment]
WIP, plan to resend patches within this month.

**Comment by jjscheel** [comment]
Thanks, @pz9115!  Please list PRs when submitted!

**Comment by pz9115** [comment]
Discussed in the last RISC-V GNU toolchain meeting, WIP.

**Comment by pz9115** [comment]
Resent patches to upstream, status change to in review, expect to generate many new discussions:

https://gcc.gnu.org/pipermail/gcc-patches/2023-November/637490.html

**Comment by pz9115** [comment]
No new recovery has been received yet， waiting for comments, maybe will discussing next SIG-Toolchain meeting in next Monday.

**Comment by pz9115** [comment]
Reviewed by Jeff Law and resend the patch, re-reviewing

https://gcc.gnu.org/pipermail/gcc-patches/2023-December/640324.html

**Comment by jjscheel** [comment]
@pz9115, can you provide an update here too?

**Comment by pz9115** [comment]
Palmer suggested to wait Profiles spec get more progress:

https://gcc.gnu.org/pipermail/gcc-patches/2023-December/641025.html

I will resend dummy extensions support in gcc part first as Christoph's suggestion:
https://gcc.gnu.org/pipermail/gcc-patches/2023-December/640332.html

**Comment by pz9115** [comment]
Still work in progress.

**Comment by pz9115** [comment]
GCC 14 was released, now preparing patches to upstream.

**Comment by pz9115** [comment]
> @pz9115, when do you expect to rebase and resend the GCC patches?

Yes, will resent it within this week.

**Comment by jjscheel** [comment]
Thanks, @pz9115.  Please post a link here with the RFC for RVA23 and RVB23.  We need that to Freeze for the new profile.

**Comment by cmuellner** [comment]
The RVI20/RVA20 support patch was resent a few hours ago: https://gcc.gnu.org/pipermail/gcc-patches/2024-July/658082.html

**Comment by pz9115** [comment]
Update RVA/B23 patch in https://gcc.gnu.org/pipermail/gcc-patches/2024-July/658502.html @jjscheel

**Comment by pz9115** [comment]
Update Binutils support in https://sourceware.org/pipermail/binutils/2024-August/136129.html

**Comment by cmuellner** [comment]
@pz9115, thanks for working on this! Your patches unblocked the Zimop/Zcmop situation, and this extension is now finally supported in Binutils:
* https://sourceware.org/git?p=binutils-gdb.git;a=commit;h=3ba06284d2cada7a4798f7849da113661aff07dd
* https://sourceware.org/git?p=binutils-gdb.git;a=commit;h=8a3ffa71945816b941d3dea9ebc7013d5be7a1a4

So, we have patches for almost everything on the lists (I think I have not seen an RFC patch for RVA23/RVB23, but that's not a big deal).

The overall strategy is now:
1. get Zimop/Zcmop into Binutils
2. get RVA20/RVA22 into Binutils
3. get Zimop/Zcmop into GCC
4. get RVA20/RVA22 into GCC
5. have RFC patches for RVA23/RVB23 for Binutils and GCC on the lists
6. wait for RVA23/RVB23 ratification
7. rebase and retest RVA23/RVB23 patches and resend them
8. get RVA23/RVB23 into Binutils
9. get RVA23/RVB23 into GCC

**Comment by pz9115** [comment]
Update the gcc patch support https://gcc.gnu.org/pipermail/gcc-patches/2024-August/660838.html

**Comment by pz9115** [comment]
Not sure if it is this doc https://lists.riscv.org/g/tech-p-ext/files/Drafts/RVP-baseInstrs-A-010.pdf,
Need to check it up then we will start the work.

**Comment by pz9115** [comment]
Received comment form Palmer and Nelson, preparing a new version patch.

https://gcc.gnu.org/pipermail/gcc-patches/2024-August/660947.html
https://sourceware.org/pipermail/binutils/2024-September/136589.html

**Comment by cmuellner** [comment]
We discussed profile support in GCC in the RISC-V BoF session at the GNU Cauldron 2024 in Prague three days ago, and we (Jeff Law, Kito, Palmer, and Jiawei were among the attendees) **reached the consensus that we will add GCC support for ratified profiles**.

There are concerns that CPU/SoC vendors will claim profile compliance when, in fact, they are not (independent of whether this happens intentionally or not). It could be as little as a tiny bug that makes a CPU/SoC incompatible with a profile it is marketing for. This could (or likely will) lead to bug reports where users blame SW implementations as being erroneous because they won't run on their device (which claims to be profile compatible). However, there is little we can do on the SW side besides closing such tickets as invalid and hoping that volunteers (in the best case, the CPU/SoC vendors themselves) engage with the upstream community to solve the problem.

We all agreed that we would like stricter compliance tests for vendors to increase the confidence level of the claims to be compatible with RVI extensions or profiles.

**Comment by pz9115** [comment]
RVA23/RVB23 going to ratification-ready candidate status, waiting for forzen

https://github.com/riscv/riscv-profiles/commit/531f859f06049725ceac77103eb980b9f4a89848

**Comment by jjscheel** [comment]
RVA23 and RVB23 are frozen and almost Ratified.  The TSC and BoD are voting:
https://docs.google.com/spreadsheets/d/1dheo5jbTkdzDtP6I0udAoYEf8i2UteSLCLN9261RhEE/edit?gid=597623560#gid=597623560

**Comment by pz9115** [comment]
Rebased and resent again to gcc upstream, reviewing
https://gcc.gnu.org/pipermail/gcc-patches/2024-December/670667.html

**Comment by jjscheel** [comment]
@pz9115 per your question about RVA23/RVB23 ratification, the [RISC-V Technical Specifications](https://lf-riscv.atlassian.net/wiki/spaces/HOME/pages/16154769/RISC-V+Technical+Specifications) wiki page reflects the proper profile documents and links to the PDFs.  Both **are** ratified.

Please let me know if anyone has more questions about this.

**Comment by pz9115** [comment]
Discussing a universal form for Profiles that should implements in compiler at first, now will follow LLVM format since it was merged into upstream, will send a new patch follow this form after the conversion discussing, https://github.com/riscv-non-isa/riscv-toolchain-conventions/pull/36

**Comment by pz9115** [comment]
Updated RISC-V Profiles CC document, waiting for merge: https://github.com/riscv-non-isa/riscv-toolchain-conventions/pull/36

Resend both gcc and binutils patches to upstream again, waiting for next round reviewing.

gcc: https://gcc.gnu.org/pipermail/gcc-patches/2025-January/674147.html
binutils: https://sourceware.org/pipermail/binutils/2025-January/138911.html

**Comment by pz9115** [comment]
Fixed BNF format, still wait to merge https://github.com/riscv-non-isa/riscv-toolchain-conventions/pull/36

**Comment by pz9115** [comment]
No new update here, still waiting for merge.

**Comment by pz9115** [comment]
No news here.

**Comment by cmuellner** [comment]
Big news here:
* LLVM patches landed (long ago): https://github.com/llvm/llvm-project/pull/76357
* the toolchain conventions PR was merged yesterday: https://github.com/riscv-non-isa/riscv-toolchain-conventions/pull/36
* the GCC patches landed (a few days ago): https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=43b450e3f72a53c744e77f55393962f1d349373a

The remaining item is Binutils, where v3 patches are on the list:
* https://sourceware.org/pipermail/binutils/2025-May/141021.html
* https://sourceware.org/pipermail/binutils/2025-May/141020.html

**Comment by pz9115** [comment]
Binutils patch get merged into upstream:
https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=3d7fb9fa5c667c0bf8760dd0db5fee7bb3e05379
https://sourceware.org/git/?p=binutils-gdb.git;a=commit;h=a3d6596ecf1cac1f81f121cb23a815895ae70645

**Comment by cmuellner** [comment]
All tasks of this ticket are completed (landed upstream).
@pz9115, thank you for the work and especially your patience and persistence.
@jjscheel, I propose to close this issue.

**Comment by gsterlin** [comment]
Thanks for the update.  It looks good and closing the ticket.
